python/mxnet/metric.py

```diff
@@ -475,8 +475,84 @@ def update(self, labels, preds):
             self.num_inst += num_samples
 
 
+class _BinaryClassificationMixin(object):
+    """
+    Private mixin for keeping track of TPR, FPR, TNR, FNR counts for a classification metric.
+    """
+
+    def __init__(self):
+        self._true_positives = 0
+        self._false_negatives = 0
+        self._false_positives = 0
+        self._true_negatives = 0
+
+    def _update_binary_stats(self, label, pred):
+        """
+        Update various binary classification counts for a single (label, pred)
+        pair.
+
+        Parameters
+        ----------
+        label : `NDArray`
+            The labels of the data.
+
+        pred : `NDArray`
+            Predicted values.
+        """
+        pred = pred.asnumpy()
+        label = label.asnumpy().astype('int32')
+        pred_label = numpy.argmax(pred, axis=1)
+
+        check_label_shapes(label, pred)
+        if len(numpy.unique(label)) > 2:
+            raise ValueError("%s currently only supports binary classification."
+                             % self.__class__.__name__)
+
+        for y_pred, y_true in zip(pred_label, label):
+            if y_pred == 1 and y_true == 1:
+                self._true_positives += 1.
+            elif y_pred == 1 and y_true == 0:
+                self._false_positives += 1.
+            elif y_pred == 0 and y_true == 1:
+                self._false_negatives += 1.
+            else:
+                self._true_negatives += 1.
+
+
+    @property
+    def _precision(self):
+        if self._true_positives + self._false_positives > 0:
+            return self._true_positives / (self._true_positives + self._false_positives)
+        else:
+            return 0.
+
+    @property
+    def _recall(self):
+        if self._true_positives + self._false_negatives > 0:
+            return self._true_positives / (self._true_positives + self._false_negatives)
+        else:
+            return 0.
+
+    @property
+    def _fscore(self):
+        if self._precision + self._recall > 0:
+            return 2 * self._precision * self._recall / (self._precision + self._recall)
+        else:
+            return 0.
+
+    @property
+    def _total_examples(self):
+        return self._false_negatives + self._false_positives + \
+               self._true_negatives + self._true_positives
+
+    def _reset_stats(self):
+        self._false_positives = 0
+        self._false_negatives = 0
+        self._true_positives = 0
+        self._true_negatives = 0
+
 @register
-class F1(EvalMetric):
+class F1(EvalMetric, _BinaryClassificationMixin):
```

Comment:
use member instead of multiple inheritance if possible

Reply:
This is one of the cases where usage of mix-ins is proper for backward compatibility and ease of use.

Reply:
I think we can make a `BinaryClassificationMetric` class that is intended to be abstract and inherits from `EvalMetric` without affecting backward compatibility. It seemed more appropriate as a mixin here since it would be useless as a concrete class and doesn't actually implement any functionality required by `EvalMetric`.

Reply:
multiple inheritance is very rarely necessary.
In this case I think _BinaryClassificationMixin should either inherit EvalMetric or be refactored into a few utility functions

Reply:
composition is also possible

Reply:
Honestly, the current multiple inheritance seems reasonable, in that calculating counts and keeping counts are two separate concerns. Mixins is more flexible and likely require less code when we extend these to multi-class/multi-label/top-k use cases.

Reply:
Composition is equally flexible with the drawback of extra dereference, though I'm fine either way.

Reply:
Can you check latest commit to see if that's what you had in mind?